src/architecture/unixfs.md

```diff
@@ -0,0 +1,611 @@
+  limits the kind of optimizations that are possible. For example, many files are
+  under the 256 KiB block size limit, so we tend to inline them into the describing
+  UnixFS `File` node. This would not be possible with an intermediate `Metadata` node.
+2. The `File` node already contains some metadata (e.g. the file size), so metadata
+  would be stored in multiple places. This complicates forwards compatibility with
+  UnixFSv2, as mapping between metadata formats potentially requires multiple fetch
+  operations.
+
+#### Metadata in the Directory
+
+Repeated `Metadata` messages are added to UnixFS `Directory` and `HAMTShard` nodes,
+the index of which indicates which entry they are to be applied to. Where entries are
+`HAMTShard`s, an empty message is added.
+
+One advantage of this method is that, if we expand stored metadata to include entry
+types and sizes, we can perform directory listings without needing to fetch further
+entry nodes (excepting `HAMTShard` nodes). However, without removing the storage of
+these datums elsewhere in the spec, we run the risk of having non-canonical data
+locations and perhaps conflicting data as we traverse through trees containing
+both UnixFS v1 and v1.5 nodes.
+
+This was rejected for the following reasons:
+
+1. When creating a UnixFS node, there's no way to record metadata without wrapping
+  it in a directory.
+2. If you access any UnixFS node directly by its [CID], there is no way of recreating
+   the metadata which limits flexibility.
+3. In order to list the contents of a directory including entry types and sizes,
+   you have to fetch the root node of each entry, so the performance benefit
+   of including some metadata in the containing directory is negligible in this
+   use case.
+
+#### Metadata in the File
+
+This adds new fields to the UnixFS `Data` message to represent the various metadata fields.
+
+It has the advantage of being simple to implement. Metadata is maintained whether
+the file is accessed directly via its [CID] or via an IPFS path that includes a
+containing directory. In addition, metadata is kept small enough that we can inline root
+UnixFS nodes into their CIDs so that we can end up fetching the same number of nodes if
+we decide to keep file data in a leaf node for deduplication reasons.
+
+Downsides to this approach are:
+
+1. Two users adding the same file to IPFS at different times will have different
+  [CID]s due to the `mtime`s being different. If the content is stored in another
+  node, its [CID] will be constant between the two users, but you can't navigate 
+  to it unless you have the parent node, which will be less available due to the 
+  proliferation of [CID]s.
+1. Metadata is also impossible to remove without changing the [CID], so 
+  metadata becomes part of the content.
+2. Performance may be impacted as well as if we don't inline UnixFS root nodes
+  into [CID]s, so additional fetches will be required to load a given UnixFS entry.
+
```

Comment:
Has "Metadata in the File" been rejected? Seems like downside 1 is significant.

Also, it would be more expected IMO if CID calculation was similar to file has calculation, in that the same hash of the file data is calculated regardless of the file's more or mtime.